src/Infrastructure/Service/Precondition/AbstractLinkIteratingPrecondition.php

```diff
@@ -0,0 +1,122 @@
+use PhpTuf\ComposerStager\Infrastructure\Value\PathList\PathList;
+
+abstract class AbstractLinkIteratingPrecondition extends AbstractPrecondition
+{
+    protected string $defaultUnfulfilledStatusMessage;
+
+    protected RecursiveFileFinderInterface $fileFinder;
+
+    protected FilesystemInterface $filesystem;
+
+    protected PathFactoryInterface $pathFactory;
+
+    final protected function getUnfulfilledStatusMessage(): string
+    {
+        return $this->defaultUnfulfilledStatusMessage;
+    }
+
+    abstract protected function getDefaultUnfulfilledStatusMessage(): string;
+
+    /** @throws \PhpTuf\ComposerStager\Domain\Exception\IOException */
+    abstract protected function isSupportedLink(PathInterface $file, PathInterface $directory): bool;
+
+    public function __construct(
+        RecursiveFileFinderInterface $fileFinder,
+        FilesystemInterface $filesystem,
+        PathFactoryInterface $pathFactory
+    ) {
+        $this->fileFinder = $fileFinder;
+        $this->filesystem = $filesystem;
+        $this->pathFactory = $pathFactory;
+        $this->defaultUnfulfilledStatusMessage = $this->getDefaultUnfulfilledStatusMessage();
+    }
+
+    public function isFulfilled(
+        PathInterface $activeDir,
+        PathInterface $stagingDir,
+        ?PathListInterface $exclusions = null
+    ): bool {
+        try {
+            $exclusions ??= new PathList([]);
+            $exclusions->add([$stagingDir->resolve()]);
+
+            if ($this->exitEarly($activeDir, $stagingDir, $exclusions)) {
+                return true;
+            }
+
+            $directories = [
+                'active' => $activeDir,
+                'staging' => $stagingDir,
+            ];
+
+            foreach ($directories as $name => $path) {
+                $files = $this->findFiles($path, $exclusions);
+
+                foreach ($files as $file) {
+                    $file = $this->pathFactory::create($file);
+
+                    if (!$this->filesystem->isLink($file)) {
+                        continue;
+                    }
+
+                    if (!$this->isSupportedLink($file, $path)) {
+                        $this->defaultUnfulfilledStatusMessage = sprintf(
+                            $this->defaultUnfulfilledStatusMessage,
+                            $name,
+                            $path->resolve(),
+                            $file->resolve(),
+                        );
+
+                        return false;
+                    }
+                }
+            }
+        } catch (InvalidArgumentException|IOException $e) {
+            // If something goes wrong, don't throw an exception--just consider the precondition
+            // unfulfilled and pass details along to the user via the status message.
+            // @todo Find a way to bubble this exception up as the $previous
+            //   argument to PreconditionException in ::assertIsFulfilled().
```

Comment:
All that's lost here is having the original exception object available via `PreconditionException::getPrevious()`. The original _message_, as you can see, _is_ captured. So if you're thinking of the end user as a Drupal site owner running automatic updates, probably no meaningful impact. They'll see the same log messages either way. If you're thinking of a developer, it will mean less access to backtrace data for debugging. Does that concern you? Should we create a follow-up issue to address it?

Reply:
Thanks for explaining the impact with such clarity & detail! 👍 

I think not taking further steps is fine. This `@todo` is sufficient IMHO. Because it won't be too difficult for a developer anyway:
- they see a message they want to understand the origin for
- they notice it comes from `isFulfilled()` failing
- the message appears nowhere in the body of `isFulfilled()`
- therefore it must come from the _only_ `catch` statement
- hence they should just put a breakpoint _here_

That seems entirely reasonable. 👍